### PR TITLE
RESP2 and RESP3 response unification

### PR DESCRIPTION
**Pull request by petyaslavova**
## RESP2/RESP3 Response Unification

This PR unifies the return types of Redis commands across RESP2 and RESP3 protocols, ensuring that application code receives consistent data structures regardless of which protocol version the client is connected with. Previously, many commands returned fundamentally different types — tuples vs lists, bytes vs strings, flat arrays vs nested structures — forcing users to write protocol-aware branching logic. After this change, RESP2 callbacks normalize responses to match RESP3 native types, making the protocol version an internal transport detail rather than a leaky abstraction.

### Core Commands
Sorted set commands (`ZRANGE`, `ZPOPMAX`, `ZSCORE`, `ZRANDMEMBER`, `ZRANK`, `ZSCAN`, etc.) now consistently return `[member, score]` lists with `float` scores, and `score_cast_func` receives a `float` in both protocols instead of raw bytes in RESP2. Blocking pop commands (`BLPOP`, `BRPOP`, `BZPOPMAX`, `BZPOPMIN`) return lists instead of tuples. `ZMPOP`/`BZMPOP` scores are always `float`. `HRANDFIELD` with values returns `[[field, value], ...]` lists in both protocols. `LCS` with `IDX` returns a dict in RESP2, matching RESP3's native map. `str_if_bytes` wrapping has been removed from RESP2 callbacks where RESP3 already returns raw types, eliminating unnecessary encoding discrepancies.

### Structural & Module Commands
Stream commands (`XINFO STREAM`, `XREAD`, `XREADGROUP`), geo commands (`GEOSEARCH`, `GEORADIUS`), and client tracking info now return matching structures across protocols. Sentinel commands normalize `flags` to `set` and masters to consistent dict shapes. Complex introspection commands (`COMMAND`, `COMMAND INFO`, `ACL LOG`, `ACL GETUSER`, `MEMORY STATS`, `DEBUG OBJECT`) received dedicated RESP3 parsers that produce the same shape as their RESP2 counterparts. Module commands across Bloom/Cuckoo/CMS/TopK/TDigest, TimeSeries, JSON, and Search all follow the same pattern — RESP3 parsers were added or aligned to produce identical output types.

### Search Module Refactoring
The Search module received significant internal cleanup beyond response unification. A shared `decode_field_value` utility in `redis/utils.py` replaces five duplicated decode-or-preserve-bytes blocks across `Result`, hybrid parsers, and `load_document`. Callback registration was deduplicated by moving `_RESP2_MODULE_CALLBACKS` / `_RESP3_MODULE_CALLBACKS` construction into a shared `_init_module_callbacks()` method on `SearchCommands`, eliminating three identical 20-line blocks. The pipeline adapter callbacks (`_search_response_callback`, `_aggregate_response_callback`, `_hybrid_response_callback`) were removed entirely — `_parse_results` now handles pipeline options directly via `functools.partial`, with a guard clause for commands queued without a query object.

### Testing
The test suite was substantially refactored to remove protocol-conditional assertions. Over 2,500 lines of duplicated `@pytest.mark.parametrize("protocol", [2, 3])` branches and `if protocol == 3:` blocks were replaced with unified assertions that validate both protocols produce the same types. New tests cover `score_cast_func` with scientific notation, `load_document` with `field_encodings`, and edge cases across all affected command families. The net result is 42 files changed with ~2,600 lines removed, reflecting the elimination of both production-side protocol branching and test-side conditional logic.

This change includes a fix for issue #4021

---

> [!NOTE]
> **High Risk**
> High risk because it introduces widespread breaking return-type/structure changes across many core and module commands (e.g., tuples→lists, bytes→str, lists→dicts, flag lists→sets) and updates pipeline callback behavior, which can break downstream code and tests that relied on protocol-specific shapes.
> 
> **Overview**
> Unifies RESP2 and RESP3 command responses so callers receive the same Python types/structures regardless of protocol, including **sorted-set score pairs** (now `list[list[member, float]]` with `score_cast_func` always receiving `float`), **blocking pops** (now lists), **streams** (`XREAD`/`XREADGROUP` now dicts), and other structural normalizations (e.g., `LCS idx` → dict, `CLIENT TRACKINGINFO` → dict, `COMMAND.flags` → set with optional `acl_categories`, Sentinel `flags` → set and RESP3 `SENTINEL MASTERS` → dict keyed by name).
> 
> Adds/updates parsers and callbacks for RESP3 parity (e.g., `MEMORY STATS`, `ACL LOG`, `DEBUG OBJECT`, `CLUSTER LINKS`, `CLUSTER SHARDS` key decoding) and normalizes geo coordinate outputs to lists.
> 
> Extends the same unification to modules: Bloom/TopK/TDigest `INFO` commands now always return `*Info` objects; fixes incorrect numeric coercion for `TOPK.ADD/INCRBY/LIST`; JSON normalizes `NUMINCRBY/NUMMULTBY` to list results, converts `JSON.RESP` float strings to native floats, and aligns `JSON.TYPE` missing-key handling; Search adds RESP3 parsing to return rich result objects and refactors module pipeline callback registration so module pipelines inherit full `client.response_callbacks` and post-process consistently.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit e1de69dd6432c01c5f2494779050e9c3a9dbc86e. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: redis/_parsers/helpers.py [hunk 1]
```diff
@@ -783,6 +950,23 @@ def parse_gcra(response, **options):
...
+    """
+    result = []
+    for lib_flat in response:
+        lib_dict = pairs_to_dict(lib_flat)
+        # Convert each function's flat list to a dict
+        if b"functions" in lib_dict:
+            lib_dict[b"functions"] = [
+                pairs_to_dict(func) for func in lib_dict[b"functions"]
+            ]
+        result.append(lib_dict)
```
[the remark this answers is not in the document]
Comment (petyaslavova) [reply redis/_parsers/helpers.py:987]: fixed.

File: redis/_parsers/helpers.py [hunk 2]
```diff
@@ -360,8 +434,8 @@ def parse_xinfo_stream(response, **options):
 
 def parse_xread(response, **options):
     if response is None:
-        return []
-    return [[r[0], parse_stream_list(r[1], **options)] for r in response]
+        return {}
+    return {r[0]: [parse_stream_list(r[1], **options)] for r in response}
```
[the remark this answers is not in the document]
Comment (petyaslavova) [reply redis/_parsers/helpers.py]: fixed.

File: redis/_parsers/helpers.py [hunk 3]
```diff
@@ -724,6 +869,28 @@ def parse_acl_log(response, **options):
...
+        if "age-seconds" in log_data:
+            log_data["age-seconds"] = float(log_data["age-seconds"])
+        # Parse client-info string into dict
+        if "client-info" in log_data:
+            log_data["client-info"] = parse_client_info(log_data["client-info"])
+        # Decode remaining string values
+        for key in log_data:
+            if key not in ("age-seconds", "client-info"):
+                log_data[key] = str_if_bytes(log_data[key])
+        data.append(log_data)
```
[the remark this answers is not in the document]
Comment (petyaslavova) [reply redis/_parsers/helpers.py:906]: false positive

File: redis/commands/json/__init__.py [hunk 4]
```diff
@@ -100,6 +95,55 @@ def _decode(self, obj):
...
+        """
+        if isinstance(obj, list):
+            return [_JSONBase._convert_resp_floats(item) for item in obj]
+        if isinstance(obj, (str, bytes)):
+            s = obj.decode() if isinstance(obj, bytes) else obj
+            try:
+                return float(s)
+            except (ValueError, OverflowError):
+                return obj
+        return obj
```
[the remark this answers is not in the document]
Comment (petyaslavova) [reply redis/commands/json/__init__.py:116]: The ambiguity the bot identifies — that a JSON string value like "42" would be converted to float(42.0) — is inherent to the RESP2 wire protocol for JSON.RESP. RESP2 sends both JSON floats and JSON strings as bulk strings, making them indistinguishable at the protocol level. This is a fundamental limitation of RESP2 that cannot be resolved without changing Redis server behavior.
The edge case of a JSON string "42" being stored and then retrieved via JSON.RESP is real but unavoidable — RESP2 literally cannot distinguish between a JSON float 42.0 and a JSON string "42" in the wire format. RESP3 can (using the , double type), which is why RESP3 doesn't need this converter.

File: redis/commands/search/commands.py [hunk 5]
```diff
@@ -175,11 +285,207 @@ def _parse_spellcheck(self, res, **kwargs):
...
+
+        RESP3 format: {"Results": {search/aggregate result dict},
+                       "Profile": {profile information dict}}
+        RESP2 format: [search/aggregate result, profile information]
+        """
+        query = kwargs["query"]
+        # RESP3 returns a dict with "Results" and "Profile" keys.
+        # Handle both decoded (str) and raw (bytes) keys.
+        results_data = res.get("Results") or res.get(b"Results") or res.get(0)
+        profile_data = res.get("Profile") or res.get(b"Profile") or res.get(1)
```
[the remark this answers is not in the document]
Comment (petyaslavova) [reply redis/commands/search/commands.py]: fixed.

File: redis/commands/search/__init__.py [hunk 6]
```diff
@@ -190,8 +178,58 @@ class Pipeline(SearchCommands, RedisPipeline):
...
+        # pipeline callback mechanism (and future cluster pipeline support)
+        # can apply them automatically — no execute() override needed.
+        self.response_callbacks[SEARCH_CMD] = partial(self._parse_results, SEARCH_CMD)
+        self.response_callbacks[AGGREGATE_CMD] = partial(
+            self._parse_results, AGGREGATE_CMD
+        )
+        self.response_callbacks[CURSOR_CMD] = partial(
+            self._parse_results, AGGREGATE_CMD
+        )
+        self.response_callbacks[HYBRID_CMD] = partial(self._parse_results, HYBRID_CMD)
```
[the remark this answers is not in the document]
Comment (petyaslavova) [reply redis/commands/search/__init__.py]: fixed

File: redis/_parsers/helpers.py [hunk 7]
```diff
@@ -724,6 +867,28 @@ def parse_acl_log(response, **options):
...
+        # Ensure age-seconds is float
+        if "age-seconds" in log_data:
+            log_data["age-seconds"] = float(log_data["age-seconds"])
+        # Parse client-info string into dict
+        if "client-info" in log_data:
+            log_data["client-info"] = parse_client_info(log_data["client-info"])
+        # Decode remaining string values
+        for key in log_data:
+            if key not in ("age-seconds", "client-info"):
+                log_data[key] = str_if_bytes(log_data[key])
```
[the remark this answers is not in the document]
Comment (petyaslavova) [reply redis/_parsers/helpers.py:905]: According to documentation, the response is not a dict

File: redis/_parsers/helpers.py [hunk 8]
```diff
@@ -928,10 +1113,16 @@ def string_keys_to_dict(key_string, callback):
...
+    "FUNCTION LIST": parse_function_list,
     "XINFO CONSUMERS": parse_list_of_dicts,
     "XINFO GROUPS": parse_list_of_dicts,
+    "HRANDFIELD": hrandfield_pairs,
+    "LCS": parse_lcs,
     "ZADD": parse_zadd,
+    "ZMPOP": parse_zmpop,
+    "BZMPOP": parse_zmpop,
     "ZMSCORE": parse_zmscore,
+    "ZRANDMEMBER": zset_score_pairs,
```
[the remark this answers is not in the document]
Comment (petyaslavova) [reply redis/_parsers/helpers.py:1146]: False positive - this is even stated in the description of the problem - that in the end, it's all fine.

[the remark this answers is not in the document]
Comment (petyaslavova) [reply redis/_parsers/helpers.py:1146]: Same as the previous one — the bot concludes "no actual bug on further reflection." False positive, nothing to do.

File: redis/_parsers/helpers.py [hunk 9]
```diff
@@ -269,6 +285,64 @@ def zset_score_pairs_resp3(response, **options):
...
+    """
+    Parse LCS response. Without modifiers returns the raw string.
+    With LEN returns an integer. With IDX returns a dict.
+    RESP2 with IDX returns a flat list [key, val, key, val, ...]
+    which we convert to a dict to match RESP3's native dict.
+    """
+    if isinstance(response, list):
+        it = iter(response)
+        return dict(zip(it, it))
+    return response
```
[the remark this answers is not in the document]
Comment (petyaslavova) [reply redis/_parsers/helpers.py:329]: fixed.

File: redis/commands/search/__init__.py [hunk 10]
```diff
@@ -190,8 +179,60 @@ class Pipeline(SearchCommands, RedisPipeline):
...
+        self.response_callbacks[PROFILE_CMD] = partial(self._parse_results, PROFILE_CMD)
```
[the remark this answers is not in the document]
Comment (petyaslavova) [reply redis/commands/search/__init__.py]: The bot walks through the logic and concludes "This is actually fine." False positive, nothing to do.

File: redis/_parsers/helpers.py [hunk 11]
```diff
@@ -620,10 +762,12 @@ def parse_command(response, **options):
         cmd_name = str_if_bytes(command[0])
         cmd_dict["name"] = cmd_name
         cmd_dict["arity"] = int(command[1])
-        cmd_dict["flags"] = [str_if_bytes(flag) for flag in command[2]]
+        cmd_dict["flags"] = {str_if_bytes(flag) for flag in command[2]}
         cmd_dict["first_key_pos"] = command[3]
         cmd_dict["last_key_pos"] = command[4]
         cmd_dict["step_count"] = command[5]
+        if len(command) > 6:
+            cmd_dict["acl_categories"] = command[6]
```
[the remark this answers is not in the document]
Comment (petyaslavova) [reply redis/_parsers/helpers.py]: fixed.

File: redis/_parsers/helpers.py [hunk 12]
```diff
@@ -880,6 +1070,7 @@ def string_keys_to_dict(key_string, callback):
     "XINFO STREAM": parse_xinfo_stream,
     "XPENDING": parse_xpending,
     "ZSCAN": parse_zscan,
+    "LCS": parse_lcs,
```
[the remark this answers is not in the document]
Comment (petyaslavova) [reply redis/_parsers/helpers.py:1088]: fixed.

File: redis/commands/search/__init__.py [hunk 13]
```diff
@@ -190,8 +179,60 @@ class Pipeline(SearchCommands, RedisPipeline):
...
+
+    @property
+    def client(self):
+        """Return self so get_protocol_version() can access connection_pool."""
+        return self
```
[the remark this answers is not in the document]
Comment (petyaslavova) [reply redis/commands/search/__init__.py:204]: No code change is needed for this issue. The Pipeline.client property returning self works correctly because self is a redis.Redis instance via the inheritance chain.

File: redis/commands/core.py [hunk 14]
```diff
@@ -7421,7 +7408,7 @@ def zdiff(
         pieces = [len(keys), *keys]
         if withscores:
             pieces.append("WITHSCORES")
-        return self.execute_command("ZDIFF", *pieces, keys=keys)
+        return self.execute_command("ZDIFF", *pieces, keys=keys, withscores=withscores)
```
[the remark this answers is not in the document]
Comment (petyaslavova) [reply redis/commands/core.py:7411]: This is a false positive. The zdiff method was different because it builds the command inline rather than delegating to _zaggregate, so it needed the explicit withscores=withscores kwarg added to its  execute_command call. zinter and zunion both go through _zaggregate, which already has:
```
    if options.get("withscores", False):
        pieces.append(b"WITHSCORES")
    options["keys"] = keys
    return self.execute_command(*pieces, **options)
```
The withscores kwarg flows through **options into  execute_command, so the callbacks will see it. No change needed.

File: .agent/resp2_vs_resp3_command_analysis.md [hunk 15]
```diff
@@ -0,0 +1,219 @@
+# RESP2 vs RESP3 Command Response Analysis — Unified in redis-py 8.0+
+
+This document lists commands where the Python return object **previously** differed between RESP2 and RESP3 protocols, and documents how each has been **unified** in redis-py 8.0+.
```
Comment (vladvildanov) [inline .agent/resp2_vs_resp3_command_analysis.md]: Differed -> different

File: tests/test_timeseries.py [hunk 16]
```diff
@@ -1424,15 +1117,6 @@ def test_mrevrange_with_count_nan_count_all_aggregators(client):
         bucket_size_msec=1000,
         filters=["type=temperature"],
     )
-    assert_resp_response(
```
Comment (vladvildanov) [inline tests/test_timeseries.py:1427]: Is it possible to entirely remove `assert_resp_response()` method? Is it used somewhere?

Comment (petyaslavova) [reply tests/test_timeseries.py:1427]: This method was defined in tests/conftest.py file, and it is deleted (there were two similar methods that were deleted). There was one other type of test assertions branching - checking directly the protocol versions - those checks were also almost removed, but we still had to keep them for some of the callbacks loadings for mocks.

File: .agent/resp2_vs_resp3_command_analysis.md (hunk 15 above)
Comment (petyaslavova) [reply .agent/resp2_vs_resp3_command_analysis.md]: Fixed.

**Review by vladvildanov — APPROVED** [review]
